lightning/src/chain/keysinterface.rs

```diff
@@ -573,13 +692,43 @@ impl KeysInterface for KeysManager {
 	}
 
 	fn get_channel_keys(&self, _inbound: bool, channel_value_satoshis: u64) -> InMemoryChannelKeys {
+		let child_ix = self.channel_child_index.fetch_add(1, Ordering::AcqRel);
+		self.derive_channel_keys(channel_value_satoshis, child_ix as u32)
+	}
+
+	fn get_onion_rand(&self) -> (SecretKey, [u8; 32]) {
+		let mut sha = self.unique_start.clone();
+
+		let child_ix = self.session_child_index.fetch_add(1, Ordering::AcqRel);
+		let child_privkey = self.session_master_key.ckd_priv(&self.secp_ctx, ChildNumber::from_hardened_idx(child_ix as u32).expect("key space exhausted")).expect("Your RNG is busted");
+		sha.input(&child_privkey.private_key.key[..]);
+
+		let mut rng_seed = sha.clone();
+		// Not exactly the most ideal construction, but the second value will get fed into
+		// ChaCha so it is another step harder to break.
+		rng_seed.input(b"RNG Seed Salt");
+		sha.input(b"Session Key Salt");
+		(SecretKey::from_slice(&Sha256::from_engine(sha).into_inner()).expect("Your RNG is busted"),
+		Sha256::from_engine(rng_seed).into_inner())
+	}
+
+	fn get_channel_id(&self) -> [u8; 32] {
+		let mut sha = self.unique_start.clone();
+
+		let child_ix = self.channel_id_child_index.fetch_add(1, Ordering::AcqRel);
+		let child_privkey = self.channel_id_master_key.ckd_priv(&self.secp_ctx, ChildNumber::from_hardened_idx(child_ix as u32).expect("key space exhausted")).expect("Your RNG is busted");
+		sha.input(&child_privkey.private_key.key[..]);
+
+		Sha256::from_engine(sha).into_inner()
+	}
+
+	fn derive_channel_keys(&self, channel_value_satoshis: u64, hallmark: u32) -> InMemoryChannelKeys {
```

Comment:
If we're going to rely on users being able to regenerate channel keys, we can't use self.unique_start in the generation of them - if you serialize the keys and then read them back from disk they'll be different!

Also, without unique_start here I'm pretty uneasy - in general we try to protect the user from accidentally using the same keys twice on reload even if they never saved the copy after the original use. Thats the point of unique_start - after restarting you'll always get new keys even if you didn't save the old one's channel_child_index counter.

The easy solution to this is to just include unique_start in the "hallmark" - instead of a u32, which doesn't carry much entropy, make it a u64 pair - the first contains the child_index and unique_start_nanos, the second unique_start_secs.

Reply:
Okay updated while storing seed, staarting_time_nanos, starting_time_secs in KeysManager to return (chan_id | starting_time_nanos, starting_time_secs) as part of ChannelKeys.

These key state is now part of SpendableOutputDescriptor and therefore can be passed to derive_channel_keys, seed is already required to be well stored between KeyManager serialization.

Let me know if you find names compelling and documentation clear enough for user.